notation.go

```diff
@@ -63,31 +42,188 @@ type SignOptions struct {
+
+	// Error that caused the verification to fail (if it fails)
+	Error error
 }
 
 // Verifier is a generic interface for verifying an artifact.
 type Verifier interface {
-	// Verify verifies the signature and returns the verified descriptor and
-	// metadata of the signed artifact.
-	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (Descriptor, error)
+	// Verify verifies the signature blob and returns the artifact
+	// descriptor upon successful verification.
+	// If signature == nil and the verification level is not 'skip', an error
+	// will be returned.
+	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (ocispec.Descriptor, *VerificationOutcome, error)
+}
+
+// Verify performs signature verification on each of the notation supported
+// verification types (like integrity, authenticity, etc.) and return the
+// verification outcomes.
+// For more details on signature verification, see
+// https://github.com/notaryproject/notaryproject/blob/main/specs/trust-store-trust-policy.md#signature-verification
+func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (ocispec.Descriptor, []*VerificationOutcome, error) {
+	var verificationOutcomes []*VerificationOutcome
+	artifactRef := opts.ArtifactReference
+	// passing nil signature to check 'skip'
+	_, outcome, err := verifier.Verify(ctx, nil, opts)
+	if err != nil {
+		if outcome == nil {
+			return ocispec.Descriptor{}, nil, err
+		}
+	} else if outcome.VerificationLevel.Name == trustpolicy.LevelSkip.Name {
+		return ocispec.Descriptor{}, []*VerificationOutcome{outcome}, nil
+	}
+
+	// get signature manifests
+	var success bool
+
+	var targetArtifactDesc ocispec.Descriptor
+	artifactDescriptor, err := repo.Resolve(ctx, artifactRef)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, ErrorSignatureRetrievalFailed{Msg: err.Error()}
+	}
+	err = repo.ListSignatures(ctx, artifactDescriptor, func(signatureManifests []ocispec.Descriptor) error {
+		// process signatures
+		for _, sigManifestDesc := range signatureManifests {
+			// get signature envelope
+			sigBlob, sigBlobDesc, err := repo.FetchSignatureBlob(ctx, sigManifestDesc)
+			if err != nil {
+				return ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("unable to retrieve digital signature with digest %q associated with %q from the registry, error : %s", sigBlobDesc.Digest, artifactRef, err.Error())}
+			}
+			payloadArtifactDescriptor, outcome, err := verifier.Verify(ctx, sigBlob, opts)
+			if err == nil && !content.Equal(payloadArtifactDescriptor, artifactDescriptor) {
+				err = errors.New("payloadArtifactDescriptor does not match artifactDescriptor")
+			}
+			if err != nil {
+				if outcome != nil {
+					outcome.SignatureBlobDescriptor = &sigBlobDesc
+					outcome.Error = err
+					verificationOutcomes = append(verificationOutcomes, outcome)
+				}
+				// TODO: add log here to track when err != nil,
+				// but outcome == nil
+				continue
+			}
```

Comment:
Always check error and then process. Otherwise, it is very easy to write buggy code.

Reply:
```suggestion
			if err != nil {
				if outcome == nil {
					// TODO: log fatal error
					return err
				}
				verificationOutcomes = append(verificationOutcomes, outcome)
				continue
			}
			if !content.Equal(payloadArtifactDescriptor, artifactDescriptor) {
				outcome.Error = errors.New("content descriptor mismatch")
				verificationOutcomes = append(verificationOutcomes, outcome)
				continue
			}
```